File: L1S/v3/code/ocaml/src/codegen.ml
let emit_line buf fmt = Printf.ksprintf (fun s -> Buffer.add_string buf (s ^ "\n")) fmt

let get_slot func vreg =
  try Hashtbl.find func.Ir.slot_map vreg
  with Not_found -> 0

let slot_addr offset = Printf.sprintf "[rbp-%d]" offset

let load_vreg buf func vreg =
  let offset = get_slot func vreg in
  if offset = 0 && false then ()  (* would be in a register *)
  else emit_line buf "\tmov rax, %s" (slot_addr offset)

let store_rax buf func vreg =
  let offset = get_slot func vreg in
  emit_line buf "\tmov %s, rax" (slot_addr offset)

let gen_prologue buf func =
  emit_line buf "\tpush rbp";
  emit_line buf "\tmov rbp, rsp";
  if func.Ir.frame_size > 0 then
    emit_line buf "\tsub rsp, %d" func.Ir.frame_size

let gen_epilogue buf func =
  emit_line buf "\tmov rsp, rbp";
  emit_line buf "\tpop rbp";
  emit_line buf "\tret"

let gen_instr buf func instr =
  match instr with
  | Ir.Const (v, n, _) ->
    emit_line buf "\tmov rax, %Ld" n;
    store_rax buf func v
  | Ir.Add (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tadd rax, rcx";
    store_rax buf func v
  | Ir.Sub (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tsub rax, rcx";
    store_rax buf func v
  | Ir.Mul (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\timul rax, rcx";
    store_rax buf func v
  | Ir.Div (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\txor rdx, rdx";
    emit_line buf "\tidiv rcx";
    store_rax buf func v
  | Ir.Eq (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tcmp rax, rcx";
    emit_line buf "\tsete al";
    emit_line buf "\tmovzx rax, al";
    store_rax buf func v
  | Ir.Lt (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tcmp rax, rcx";
    emit_line buf "\tsetl al";
    emit_line buf "\tmovzx rax, al";
    store_rax buf func v
  | Ir.And (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tand rax, rcx";
    store_rax buf func v
  | Ir.Or (v, a, b) ->
    load_vreg buf func a;
    emit_line buf "\tpush rax";
    load_vreg buf func b;
    emit_line buf "\tmov rcx, rax";
    emit_line buf "\tpop rax";
    emit_line buf "\tor rax, rcx";
    store_rax buf func v
  | Ir.Not (v, a) ->
    load_vreg buf func a;
    emit_line buf "\tnot rax";
    store_rax buf func v
  | Ir.Neg (v, a) ->
    load_vreg buf func a;
    emit_line buf "\tneg rax";
    store_rax buf func v
  | Ir.Load (v, src, _) ->
    load_vreg buf func src;
    emit_line buf "\tmov rax, [rax]";
    store_rax buf func v
  | Ir.Store (dst, src, _) ->
    load_vreg buf func dst;
    emit_line buf "\tpush rax";
    load_vreg buf func src;
    emit_line buf "\tpop rcx";
    emit_line buf "\tmov [rcx], rax"
  | Ir.Alloc _ -> ()  (* handled by frame allocation *)
  | Ir.Call (v, name, args, _) ->
    let arg_regs = [| "rdi"; "rsi"; "rdx"; "rcx"; "r8"; "r9" |] in
    List.iteri (fun i a ->
      if i < 6 then begin
        load_vreg buf func a;
        emit_line buf "\tmov %s, rax" arg_regs.(i)
      end
    ) args;
    if List.length args > 6 then
      (* push extra args to stack in reverse *)
      let extra = List.rev (List.drop 6 args) in
      List.iter (fun a ->
        load_vreg buf func a;
        emit_line buf "\tpush rax"
      ) extra;
    emit_line buf "\tcall %s" name;
    store_rax buf func v

let gen_terminator buf func = function
  | Ir.Ret v ->
    load_vreg buf func v;
    gen_epilogue buf func
  | Ir.Br label ->
    emit_line buf "\tjmp %s" label
  | Ir.BrCond (v, t_label, f_label) ->
    load_vreg buf func v;
    emit_line buf "\ttest rax, rax";
    emit_line buf "\tjnz %s" t_label;
    emit_line buf "\tjmp %s" f_label

let gen_func buf func =
  emit_line buf "";
  emit_line buf "global %s" func.Ir.name;
  emit_line buf "%s:" func.Ir.name;
  gen_prologue buf func;
  List.iter (fun block ->
    emit_line buf "%s:" block.Ir.label;
    List.iter (fun instr -> gen_instr buf func instr) (List.rev block.Ir.instrs);
    gen_terminator buf func block.Ir.terminator
  ) func.Ir.blocks

let generate_module modl =
  let buf = Buffer.create 4096 in
  emit_line buf "default rel";
  emit_line buf "section .text";
  List.iter (fun func -> gen_func buf func) modl.Ir.funcs;
  Buffer.contents buf

File: L1S/v3/code/rust/src/codegen.rs
use crate::ir::*;
use crate::regalloc::StackAllocator;

pub struct Codegen {
    output: String,
    alloc: StackAllocator,
}

impl Codegen {
    pub fn new(alloc: StackAllocator) -> Self {
        Self { output: String::new(), alloc }
    }

    fn emit(&mut self, s: &str) {
        self.output.push_str(s);
        self.output.push('\n');
    }

    pub fn generate(&mut self, module: &IrModule) -> String {
        self.emit("section .text");
        self.emit("global main");
        self.emit("");
        for func in &module.funcs {
            self.gen_func(func);
        }
        std::mem::take(&mut self.output)
    }

    fn gen_func(&mut self, func: &IrFunction) {
        self.alloc.allocate(func.num_vregs);
        self.emit(&format!("{}:", func.name));
        self.emit("  push rbp");
        self.emit("  mov rbp, rsp");
        self.emit(&format!("  sub rsp, {}", self.alloc.frame_size));

        let arg_regs = ["rdi", "rsi", "rdx", "rcx", "r8", "r9"];
        for (i, &pv) in func.param_vregs.iter().enumerate() {
            if i < arg_regs.len() {
                self.emit(&format!("  mov [rbp-{}], {}", self.alloc.offset(pv), arg_regs[i]));
            }
        }

        for block in &func.blocks {
            self.gen_block(block, func);
        }
        self.emit("");
    }

    fn gen_block(&mut self, block: &IrBlock, _func: &IrFunction) {
        self.emit(&format!(".{}:", block.label));

        for instr in &block.instrs {
            match instr {
                IrInstr::Const(vreg, val, _) => {
                    let off = self.alloc.offset(*vreg);
                    self.emit(&format!("  mov rax, {}", val));
                    self.emit(&format!("  mov [rbp-{}], rax", off));
                }
                IrInstr::Mov(dst, src) => {
                    let d = self.alloc.offset(*dst);
                    let s = self.alloc.offset(*src);
                    self.emit(&format!("  mov rax, [rbp-{}]", s));
                    self.emit(&format!("  mov [rbp-{}], rax", d));
                }
                IrInstr::Add(dst, a, b) => self.binop("add", dst, a, b),
                IrInstr::Sub(dst, a, b) => self.binop("sub", dst, a, b),
                IrInstr::Mul(dst, a, b) => self.binop("imul", dst, a, b),
                IrInstr::Div(dst, a, b) => {
                    let da = self.alloc.offset(*a);
                    let db = self.alloc.offset(*b);
                    let dd = self.alloc.offset(*dst);
                    self.emit(&format!("  mov rax, [rbp-{}]", da));
                    self.emit("  cdq");
                    self.emit(&format!("  mov rcx, [rbp-{}]", db));
                    self.emit("  idiv rcx");
                    self.emit(&format!("  mov [rbp-{}], rax", dd));
                }
                IrInstr::Eq(dst, a, b) => self.setcc("sete", dst, a, b),
                IrInstr::Lt(dst, a, b) => self.setcc("setl", dst, a, b),
                IrInstr::And(dst, a, b) => self.binop("and", dst, a, b),
                IrInstr::Or(dst, a, b) => self.binop("or", dst, a, b),
                IrInstr::Not(dst, src) => {
                    let ds = self.alloc.offset(*src);
                    let dd = self.alloc.offset(*dst);
                    self.emit(&format!("  mov rax, [rbp-{}]", ds));
                    self.emit("  cmp rax, 0");
                    self.emit("  sete al");
                    self.emit("  movzx rax, al");
                    self.emit(&format!("  mov [rbp-{}], rax", dd));
                }
                IrInstr::Neg(dst, src) => {
                    let ds = self.alloc.offset(*src);
                    let dd = self.alloc.offset(*dst);
                    self.emit(&format!("  mov rax, [rbp-{}]", ds));
                    self.emit("  neg rax");
                    self.emit(&format!("  mov [rbp-{}], rax", dd));
                }
                IrInstr::Call(result, name, args, _) => {
                    let arg_regs = ["rdi", "rsi", "rdx", "rcx", "r8", "r9"];
                    for (i, arg) in args.iter().enumerate() {
                        if i < arg_regs.len() {
                            let off = self.alloc.offset(*arg);
                            self.emit(&format!("  mov {}, [rbp-{}]", arg_regs[i], off));
                        }
                    }
                    self.emit(&format!("  call {}", name));
                    let dr = self.alloc.offset(*result);
                    self.emit(&format!("  mov [rbp-{}], rax", dr));
                }
            }
        }

        if let Some(term) = &block.terminator {
            match term {
                IrTerminator::Ret(vreg_opt) => {
                    if let Some(vreg) = vreg_opt {
                        let off = self.alloc.offset(*vreg);
                        self.emit(&format!("  mov rax, [rbp-{}]", off));
                    }
                    self.emit("  mov rsp, rbp");
                    self.emit("  pop rbp");
                    self.emit("  ret");
                }
                IrTerminator::Br(label) => {
                    self.emit(&format!("  jmp .{}", label));
                }
                IrTerminator::BrCond(vreg, tlabel, flabel) => {
                    let off = self.alloc.offset(*vreg);
                    self.emit(&format!("  mov rax, [rbp-{}]", off));
                    self.emit("  cmp rax, 0");
                    self.emit(&format!("  je .{}", flabel));
                    self.emit(&format!("  jmp .{}", tlabel));
                }
            }
        }
    }

    fn binop(&mut self, op: &str, dst: &VReg, a: &VReg, b: &VReg) {
        let da = self.alloc.offset(*a);
        let db = self.alloc.offset(*b);
        let dd = self.alloc.offset(*dst);
        self.emit(&format!("  mov rax, [rbp-{}]", da));
        self.emit(&format!("  mov rcx, [rbp-{}]", db));
        self.emit(&format!("  {} rax, rcx", op));
        self.emit(&format!("  mov [rbp-{}], rax", dd));
    }

    fn setcc(&mut self, cc: &str, dst: &VReg, a: &VReg, b: &VReg) {
        let da = self.alloc.offset(*a);
        let db = self.alloc.offset(*b);
        let dd = self.alloc.offset(*dst);
        self.emit(&format!("  mov rax, [rbp-{}]", da));
        self.emit(&format!("  mov rcx, [rbp-{}]", db));
        self.emit("  cmp rax, rcx");
        self.emit(&format!("  {} al", cc));
        self.emit("  movzx rax, al");
        self.emit(&format!("  mov [rbp-{}], rax", dd));
    }
}
